azurelinuxagent/ga/agent_update_handler.py

```diff
@@ -160,38 +198,142 @@ def __get_agent_family_manifests(self, goal_state):
+            elif conf.get_enable_ga_versioning() and self._is_vm_enabled_for_rsm_upgrades:
+                updater = RSMVersionUpdater(gs_id, agent_family, None, version, self.update_state, self._is_version_from_rsm, self._daemon_version)
+                self.update_state.last_attempted_update_version = version
+            else:
+                updater = SelfUpdateVersionUpdater(gs_id, agent_family, None, None, self.update_state)
+
+            # verify if agent update is allowed
+            if not updater.should_update_agent(goal_state):
+                return
+            updater.log_new_agent_update_message()
+            updater.purge_extra_agents_from_disk()
+            agent = updater.download_and_get_new_agent(self._protocol, goal_state)
+            if agent.is_blacklisted or not agent.is_downloaded:
+                msg = "Downloaded agent version is in bad state : {0} , skipping agent update".format(
+                    str(agent.version))
+                raise AgentUpdateError(msg)
+            updater.proceed_with_update()
+
+        except Exception as err:
+            if isinstance(err, AgentUpgradeExitException):
+                raise err
+            elif isinstance(err, AgentUpdateError):
+                error_msg = ustr(err)
+            else:
+                error_msg = "Unable to update Agent: {0}".format(textutil.format_exception(err))
+            logger.warn(error_msg)
+            add_event(op=WALAEventOperation.AgentUpgrade, is_success=False, message=error_msg, log_event=False)
+            self.update_state.last_attempted_update_error_msg = error_msg
+
+    def get_vmagent_update_status(self):
+        """
+        This function gets the VMAgent update status as per the last attempted update.
+        Returns: None if fail to report or update never attempted with rsm version specified in GS
+        """
+        try:
+            if conf.get_enable_ga_versioning() and self._is_vm_enabled_for_rsm_upgrades and self._is_version_from_rsm:
+                if not self.update_state.last_attempted_update_error_msg:
+                    status = VMAgentUpdateStatuses.Success
+                    code = 0
+                else:
+                    status = VMAgentUpdateStatuses.Error
+                    code = 1
+                return VMAgentUpdateStatus(expected_version=str(self.update_state.last_attempted_update_version), status=status, code=code, message=self.update_state.last_attempted_update_error_msg)
+        except Exception as err:
+            msg = "Unable to report agent update status: {0}".format(textutil.format_exception(err))
+            logger.warn(msg)
+            add_event(op=WALAEventOperation.AgentUpgrade, is_success=False, message=msg, log_event=True)
+        return None
 
-    def __download_and_get_agent(self, goal_state, agent_family, agent_manifest, requested_version):
+
+class GAVersionUpdater(object):
```

Comment:
Thanks for creating those classes. It's a good step in the right direction. However, some of the state and logic that should be in these classes is still spread around other classes.

As an example, this state should be in these 2 derived classses:

```
    def __init__(self, gs_id, agent_family, agent_manifest, version, update_state):
        self._gs_id = gs_id
        self._agent_family = agent_family
        self._agent_manifest = agent_manifest
        self._version = version
        self._update_state = update_state

```

Same for these properties:

```
class AgentUpdateHandler(object):
    def __init__(self, protocol):
         ...
        # restore the state of rsm update
        if not os.path.exists(self._get_rsm_version_state_file()):
            self._is_version_from_rsm = False
            self._is_vm_enabled_for_rsm_upgrades = False
        else:
            self._is_version_from_rsm = self._get_is_version_from_rsm()
            self._is_vm_enabled_for_rsm_upgrades = self._get_is_vm_enabled_for_rsm_upgrades()
```

And this logic: 

```
    def _get_is_version_from_rsm(self):
    def _get_is_vm_enabled_for_rsm_upgrades(self):
    def _get_rsm_state_used_gs_timestamp(self):
    def _update_rsm_version_state_if_changed(self, goalstate_timestamp, agent_family):
```

At this point you can't move this state to the updaters because you are creating them on each iteration of the main loop and this state needs to persist across iterations. But instantiating on each iteration is not needed. You only need to instantiate them only when there is a new goal state and the "is vm enabled for RSM updates" property changes. This approach would also address my other concern about having too much code/logic executing on each iteration of the main loop.